### ISSUE TITLE
Use MIOpen Optimized RNN layers

### ISSUE THREAD

**Issue by Delaunay** [ROCmSoftwarePlatform/pytorch#388]
## 🚀 Feature

At the moment rocm pytorch RNN cells use aten GPU code to run RNN and not [MIOpen optimized implementation][1].

[1]: https://github.com/ROCmSoftwarePlatform/MIOpen/blob/6c130927bc5053eb4290dd937ecd1514d3283c00/include/miopen/miopen.h#L2191

## Motivation

* RNN is significantly slower on AMD (x3 on my hardware, RX580 vs 1060)

**Comment by iotamudelta** [comment]
@Delaunay to update here, I've also recently looked at the aten implementation and found quite some time spend in matrix multiplications. So tuning as for #398 will apply.

**Comment by clemej** [comment]
This may be related, so posting here.

I'm seeing ~10x slower training on the exact same python code using LSTMs on VEGA56 versus a GTX 1070.  rocm-smi reports GPU being used, but clearly not optimally.  Using self-compiled pytorch v1.1.0 from pytorch main repo (not here, should I be using this repo?)  with ROCm 2.5.27 on 18.02.4  .  These are small networks (1x1 input layer, 2x40 LSTM layer, and 1x2 output layer).  Dataset take ~30s per-training epoch on VEGA56, only approx. ~3 seconds on 1070. 

On a separate non-RNN experiment using a large but simple feed forward network, performance is quite good compared to the 1070.

Compiled with `PYTORCH_ROCM_ARCH=gfx900 USE_ROCM=1 USE_LMDB=1 USE_OPENCV=1 MAX_JOBS=8 BLAS="OpenBLAS" USE_MKLDNN=0 USE_MIOPEN=1 python3 setup.py bdist_wheel` .  

Code also prints out spurious CUDA warning about pytorch not being compiled with cuDNN at the beginning.. which of course it isn't. `PyTorch was compiled without cuDNN support. To use cuDNN, rebuild `

**Comment by Delaunay** [comment]
It is related. Optimized RNN layers include:
* LSTM
* GRU
* RNN_RELU
* RNN_TANH

So it makes sense you see that kind of performance.

**Comment by clemej** [comment]
I figured. I see the discussion on #398 but I'm running 2.5 and that doesn't seem to help. Any branches I can try?

**Comment by iotamudelta** [comment]
MIOpen RNN integration to PyTorch was merged in 9f8e2c0 - closing for now.

**Comment by Delaunay** [comment]
I did some synthetic benchmarks and I am able to get 3x on some configurations.
Although, I having trouble making an actual model using the MIOpen path for RNNs.

I tried to make [wlm][1] used the optimized kernel by simply removing the dropout (`dropout=0`)
I think as long as `nlayers > 1` and that the tensors are floats it should use the MIOpen implementation but I am not able to get it to work.  

[1]: https://github.com/pytorch/examples/blob/master/word_language_model/model.py
